x-pack/plugins/alerting/server/task_runner/execution_handler.ts

```diff
@@ -0,0 +1,418 @@
+            logger.debug(
+              `Rule "${this.rule.id}" skipped scheduling action "${action.id}" because the maximum number of allowed actions for connector type ${actionTypeId} has been reached.`
+            );
+          }
+          ruleRunMetricsStore.setTriggeredActionsStatusByConnectorType({
+            actionTypeId,
+            status: ActionsCompletion.PARTIAL,
+          });
+          continue;
+        }
+
+        if (!this.isActionExecutable(action)) {
+          this.logger.warn(
+            `Rule "${this.taskInstance.params.alertId}" skipped scheduling action "${action.id}" because it is disabled`
+          );
+          continue;
+        }
+
+        ruleRunMetricsStore.incrementNumberOfTriggeredActions();
+        ruleRunMetricsStore.incrementNumberOfTriggeredActionsByConnectorType(actionTypeId);
+
+        const actionToRun = {
+          ...action,
+          params: injectActionParams({
+            ruleId,
+            spaceId,
+            actionTypeId,
+            actionParams: transformActionParams({
+              actionsPlugin,
+              alertId: ruleId,
+              alertType: this.ruleType.id,
+              actionTypeId,
+              alertName: this.rule.name,
+              spaceId,
+              tags: this.rule.tags,
+              alertInstanceId: alertId,
+              alertActionGroup: actionGroup,
+              alertActionGroupName: this.ruleTypeActionGroups.get(actionGroup)!,
+              context: alert.getContext(),
+              actionId: action.id,
+              state,
+              kibanaBaseUrl: this.taskRunnerContext.kibanaBaseUrl,
+              alertParams: this.rule.params,
+              actionParams: action.params,
+            }),
+          }),
+        };
+
+        await this.actionRunOrAddToBulk({
+          enqueueOptions: this.getEnqueueOptions(actionToRun),
+          actionsClient,
+          bulkActions,
+        });
+
+        logActions.push({
+          id: action.id,
+          typeId: action.actionTypeId,
+          alertId,
+          alertGroup: action.group,
+        });
+
+        if (recovered) {
+          alert.scheduleActions(action.group as ActionGroupIds);
+        }
+      }
+
+      if (!!bulkActions.length) {
+        for (const c of chunk(bulkActions, CHUNK_SIZE)) {
+          await actionsClient.bulkEnqueueExecution(c);
+        }
+      }
+
+      if (!!logActions.length) {
+        for (const action of logActions) {
+          alertingEventLogger.logAction(action);
+        }
+      }
+    }
+  }
+
+  private generateExecutables({
+    actions,
+    alerts,
+    recovered,
+  }: {
+    actions: RuleAction[];
+    alerts: Record<string, Alert<State, Context, ActionGroupIds | RecoveryActionGroupId>>;
+    recovered: boolean;
+  }) {
+    const executables = [];
+
+    for (const action of actions) {
+      for (const [alertId, alert] of Object.entries(alerts)) {
```

Comment:
> Could we make these loops more efficient by only checking if `this.isAlertExecutable` if the current `action` matches the scheduled action group? It seems like we're doing a lot of unnecessary checks right now.

Actually this was a draft and my goal was just changing the alert > action execution order without breaking any test and then getting feedbacks about the possible pitfalls such as `alert.updateLastScheduledActions(action.group as ActionGroupIds);`. But as i didn't get response and i was away, moved it to review column :)

Changing `this.isAlertExecutable` is a good idea but probably would break a lot of test cases.
I think we can do that once we are sure that every thing is working as expected.

i am fixing the other nits :)

Reply:
I tried to do this but didn't work... 
We can get the actionGroup and state only after checking if an alert is executable. So `this.isAlertExecutable` has to be before `if (action.group === actionGroup)`

Reply:
Done :)